jax/_src/lax/lax.py

```diff
@@ -1527,9 +1528,14 @@ def broadcast_mhlo(
       assert len(aval.shape) <= len(aval_out.shape), (aval, aval_out)
       dims = mlir.dense_int_elements(
           range(len(aval_out.shape) - len(aval.shape), len(aval_out.shape)))
-      arg = mhlo.BroadcastInDimOp(
-          mlir.aval_to_ir_type(aval.update(shape=aval_out.shape)), arg,
-          dims).result
+      if any(isinstance(d, ir.Value) for d in aval_out.shape):
```

Comment:
At the moment, applications like `add x:f32[] y:f32[n]` are legal in dynamic shape jaxprs. This `broadcast_mhlo` function was written to handle things like `add x:f32[] y:f32[3]` by calling `BroadcastInDimOp`, but to handle cases like `add x:f32[] y:f32[n]` we had to extend it to sometimes call `DynamicBroadcastInDimOp`.